### PR TITLE
Bump tree-sitter-r 5 - Leading newline skipping, tree-sitter 0.23.0 update

### PR DESCRIPTION
**Pull request by DavisVaughan**
Contains these commits https://github.com/r-lib/tree-sitter-r/compare/99bf614d9d7e6ac9c7445fa7dc54a590fcdf3ce0...2097fa502efa21349d26af0ffee55d773015e481, most importantly:
- https://github.com/r-lib/tree-sitter-r/pull/134 to fix some panics we'd see in `Document::new()` when there are leading newlines before the first actual token in a document - I can't find a linked issue for it, but I added a test, and there are many on the tree-sitter-r side
- https://github.com/r-lib/tree-sitter-r/issues/141 to fix an edge case with if/else and newlines
- https://github.com/r-lib/tree-sitter-r/pull/145 to fix an edge case with functions and newlines

### CODE REVIEW

File: crates/ark/src/lsp/documents.rs [hunk 1]
```diff
@@ -224,4 +225,11 @@ mod tests {
         let point = compute_point(Point::new(0, 0), "abcdefghi\n");
         assert_eq!(point, Point::new(1, 0));
     }
+
+    #[test]
+    fn test_document_starts_at_0_0_with_leading_whitespace() {
+        let document = Document::new("\n\n# hi there", None);
+        let root = document.ast.root_node();
+        assert_eq!(root.start_position(), Point::new(0, 0));
+    }
```
Comment (DavisVaughan) [inline crates/ark/src/lsp/documents.rs:234]: This is the test that ensures that the `program` node should start at `(0, 0)` and cover the whole document.

Comment (DavisVaughan) [reply crates/ark/src/lsp/documents.rs:234]: It was previously possible for the program node to start at, say, `(3, 0)` if thats where the first content of the document is. Then we'd request something like "find closest node to" `(0, 1)` and that in theory should return the `program` node as the top level node which covers the whole document, but since it started at `(3, 0)` we'd get a panic (or at least an error log).

This isn't every source of these errors, but it was one I saw a decent bit

File: crates/ark/src/lsp/snapshots/indent.R [hunk 2]
```diff
@@ -1098,11 +1098,11 @@ fun_call(object1 + object2 ~ object3 +
...
 
 ## 32
-fun_call(object ~
+fun_call(~ object
   )
 
 ## 33
-fun_call(object +
+fun_call(object + object2
   )
```
[the remark this answers is not in the document]
Comment (lionel-) [reply crates/ark/src/lsp/snapshots/indent.R:1106]: I'm not sure about all these changes to `indent.R`. Here these tests were intentionally checking invalid code. An indenter should ideally work no matter the validity of the code (i.e. partially written code as in these examples).

I guess it's fine if you want to take it as is for formatter tests, where we almost surely don't want to take on the challenge to format invalid code. We'll have to think about what to do with indentation of invalid code though. Do we all of a sudden start indenting pipelines in staircase when the file is invalid?

**Review by lionel- — APPROVED** [review]

File: crates/ark/src/lsp/snapshots/indent.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (DavisVaughan) [reply crates/ark/src/lsp/snapshots/indent.R:1106]: >  An indenter should ideally work no matter the validity of the code

But an indenter requires you to first parse the code, and if the code doesn't parse, you are going to have trouble indenting

[the remark this answers is not in the document]
Comment (DavisVaughan) [reply crates/ark/src/lsp/snapshots/indent.R:1106]: I think that an indenter and a formatter are one and the same in my head, but it does not seem like you agree

Like to me people aren't going to be applying reindentation without also reformatting. At least, I don't think our tools are going to be allowing that. ruff doesn't, and it doesn't seem like the LSP has a request specifically for indentation, just formatting.

[the remark this answers is not in the document]
Comment (lionel-) [reply crates/ark/src/lsp/snapshots/indent.R:1106]: > and if the code doesn't parse, you are going to have trouble indenting

That's why we use error-recovering parsers to power indenters (or regular expressions).

> I think that an indenter and a formatter are one and the same in my head, but it does not seem like you agree

They _can_ be, but a formatter is in general something more static than an indenter which is more dynamic. An indenter has to work no matter the state of the text because the user might be in the middle of writing code.

The two snippets you changed here are a good illustration, here is a common case the indenter needs to support:

```r
foo(bar +<cursor>)

# <enter>
foo(bar +
  <cursor>)
```

And that's a case where parentheses were auto-closed. The indenter should also know what to do with missing parentheses.

> it doesn't seem like the LSP has a request specifically for indentation, just formatting

In practice it's format-as-you-type but yes this is missing from the LSP, in part because editors treat indentation as latency-critical to provide a good experience to end users.

That said, fixing up indentation via the LSP works well enough in practice (as done in the Python extension with Pylance or positron-r with Ark).

We could split the task in two separate engines, an indenter that works on broken code and a formatter that requires valid code. In that case we would need to make sure the behaviours don't differ too much to avoid making it look like the editor is working against itself. There's also the question of the amount of work, though I think we have determined that the rule-based indenter does not require too much work.

The ideal case though is that the formatter itself can be used for indenting, so we need to think about whether it's really too expensive to let the formatter handle error-recovered code.

[the remark this answers is not in the document]
Comment (lionel-) [reply crates/ark/src/lsp/snapshots/indent.R:1106]: Another use case to think about:

Copying parts of some code, and pasting it with formatting enabled. It would be nice to at least indent the code.

[the remark this answers is not in the document]
Comment (DavisVaughan) [reply crates/ark/src/lsp/snapshots/indent.R:1106]: > We could split the task in two separate engines, an indenter that works on broken code and a formatter that requires valid code.

Honestly I think that this doesn't sound too bad. But I'd be in favor of:
- A quite stupid indenter for Format on Type that just uses some basic rules and gets it right "most" of the time (like after a pipe or after a `+` like above). I guess it's ok that it uses the tree-sitter parser even on broken code. We will have to see. We've obviously determined that _just_ regex on the frontend isn't enough for this, so we probably do need the parser.
- A good formatter that only works on valid code, hooked up to Format on Save. In the 5% of the time that the indenter falls over, you just finish out the code to make it valid and hit "save" and then it gets fixed up and you keep going.

I think this is mostly how writing Rust feels and it doesn't bother me
